editor/editor/layout.js

```diff
@@ -17,8 +17,7 @@ class Layout extends wp.element.Component {
 	}
 
 	switchMode( newMode ) {
-		// TODO: we need a serializer from blocks here
-		const html = this.state.html;
+		const html = this.mode === 'text' ? this.state.html : wp.blocks.serialize( this.state.blocks );
```

Comment:
The since-removed "props in getInitialState as anti-pattern" article spoke to this specific usage of state as a cache to often be a poor choice for the impact on maintainability caused by duplicating the source of truth:

>Using props to generate state in getInitialState often leads to duplication of "source of truth", i.e. where the real data is. This is because getInitialState is only invoked when the component is first created.
>
>Whenever possible, compute values on-the-fly to ensure that they don't get out of sync later on and cause maintenance trouble.

https://github.com/facebook/react/blob/f1c1544/docs/tips/10-props-in-getInitialState-as-anti-pattern.md

In our specific case, it also seems to assume that we don't have a good sense of when `render` will be called. As-is, we should expect it only to be called on initialization and on mode changes, which is where we're regenerating `html` and `blocks` state anyways.